src/data/nat/cast.lean

```diff
@@ -242,6 +242,21 @@ lemma map_nat_cast (f : A →+ B) (h1 : f 1 = 1) (n : ℕ) : f n = n :=
 
 end add_monoid_hom
 
+namespace monoid_with_zero_hom
+
+variables {A : Type*} [monoid_with_zero A]
+
+/-- If two `monoid_with_zero_hom`s agree on the positive naturals they are equal. -/
+@[ext] theorem ext_nat {f g : monoid_with_zero_hom ℕ A}
+  (h_pos : ∀ {n : ℕ}, 0 < n → f n = g n) : f = g :=
+begin
+  ext (_ | n),
+  { rw [f.map_zero, g.map_zero] },
+  { exact h_pos n.zero_lt_succ, },
+end
```

Comment:
We could of course go further and say:
```suggestion
/-- If two `monoid_with_zero_hom`s agree on the positive naturals they are equal. -/
@[ext] theorem ext_nat {f g : monoid_with_zero_hom ℕ A}
  (h_pos : ∀ {n : ℕ}, 2 ≤ n → f n = g n) : f = g :=
begin
  ext (_ | _ | n),
  { rw [f.map_zero, g.map_zero] },
  { rw [f.map_one, g.map_one] },
  { exact h_pos (nat.succ_le_succ $ nat.succ_le_succ n.zero_le), },
end
```

but perhaps this is best left to a follow-up that says it need only agree on the primes.

Reply:
Let me know if you would like me to make this change in this PR or leave it. I don't have any inclination of which is better.

Reply:
In fact, I'm tempted to just revert this lemma, since I can't see it really being useful

Reply:
Whatever you think is best, just let me know!

Reply:
Is there any harm in having both? (with the one that is weaker but more useful derived from the one that is stronger)